src/transform/src/reduction_pushdown.rs

```diff
@@ -81,8 +90,707 @@ impl ReductionPushdown {
+            } else if agg.distinct && agg.func.ignores_duplicates() {
+                outer_aggs.push((agg_num, agg.clone()));
+            } else if let Some(expr_input) = expr_inputs.next() {
+                if expr_inputs.next().is_some() {
+                    // an aggregation involving more than one input
+                    outer_aggs.push((agg_num, agg.clone()));
+                } else {
+                    // an aggregation involving one input.
+                    let mut localized_agg = agg.clone();
+                    localized_agg.expr = old_join_mapper.map_expr_to_local(localized_agg.expr);
+                    if let MirRelationExpr::Reduce { group_key, .. } = &inputs[expr_input] {
+                        if localized_agg
+                            .expr
+                            .support()
+                            .iter()
+                            .any(|i| i >= &group_key.len())
+                        {
+                            // skip pushing down the aggregation if it aggregates a
+                            // partial aggregate to prevent infinite layers of
+                            // partial reduction pushdown
+                            outer_aggs.push((agg_num, agg.clone()));
+                            continue;
+                        }
+                    }
+                    single_input_aggs_to_push[expr_input]
+                        .0
+                        .push((agg_num, localized_agg));
+                }
+            } else {
+                // an aggregation involving no inputs
+                if agg.func.row_count_dependent() {
+                    no_input_aggs_to_push.push((agg_num, agg.clone()));
+                } else {
+                    // It is not worthwhile to push down a reduce whose result
+                    // only depends on whether or not an input is empty. such a
+                    // reduce should be simplified to `Distinct ()` + a map
+                    outer_aggs.push((agg_num, agg.clone()));
+                }
+            }
+        }
+
+        // Classify predicates by how many inputs the predicate references
+        for predicate in predicates {
+            let mut expr_inputs = old_join_mapper.lookup_inputs(predicate);
+            if let Some(expr_input) = expr_inputs.next() {
+                if expr_inputs.next().is_some() {
+                    // TODO: support multi-input predicates
+                    // TODO: attempt to rewrite multi-input predicates in terms of a
+                    // single input
+                    return None;
+                } else {
+                    single_input_aggs_to_push[expr_input]
+                        .1
+                        .push(old_join_mapper.map_expr_to_local(predicate.clone()))
+                }
+            }
+        }
+
+        // Union together the supports of all outer aggs, which will become
+        // `aggs_not_pushed_support`.
```

Comment:
I got to here when I realized that I just didn't know what the plan was for this code, and was no longer keeping everything in my head. I think the conclusion there is that we (I, at least) need more module-level documentation outlining the "plan" for what the code is going to do. I can believe that this information is in this file, but if we can hoist it upwards so that the sequence of steps are explained, and then each can be independently understood, that would make understanding this so much easier.